File: engine/modelcraft_engine/hw.py
"""Hardware probe and adaptive training planner (machine-agnostic)."""

from __future__ import annotations

import json
import os
import re
import shutil
import subprocess
from dataclasses import asdict, dataclass, field
from pathlib import Path
from typing import Any

try:
    import psutil
except ImportError:  # pragma: no cover
    psutil = None  # type: ignore


def _subprocess_kwargs() -> dict:
    if os.name == "nt":
        return {"creationflags": getattr(subprocess, "CREATE_NO_WINDOW", 0x08000000)}
    return {}


@dataclass
class GpuInfo:
    index: int
    name: str
    vram_total_mb: int
    vram_used_mb: int
    vram_free_mb: int
    utilization_pct: float | None = None
    temperature_c: float | None = None


@dataclass
class HardwareSnapshot:
    gpus: list[GpuInfo] = field(default_factory=list)
    primary_gpu: GpuInfo | None = None
    ram_total_mb: int = 0
    ram_used_mb: int = 0
    ram_available_mb: int = 0
    cpu_count: int = 1
    cpu_usage_pct: float | None = None
    disk_free_mb: int | None = None
    disk_total_mb: int | None = None
    disk_path: str | None = None
    nvidia_smi_ok: bool = False
    cuda_ready: bool | None = None
    notes: list[str] = field(default_factory=list)


@dataclass
class TrainingPlan:
    power_mode: str
    load_mode: str
    max_seq_length: int
    batch_size: int
    grad_accum: int
    lora_rank: int
    max_steps: int
    estimated_vram_mb: int
    hard_blocks: list[str] = field(default_factory=list)
    soft_warnings: list[str] = field(default_factory=list)
    summary: str = ""


def _run_nvidia_smi() -> tuple[list[GpuInfo], bool]:
    exe = shutil.which("nvidia-smi")
    if not exe:
        return [], False
    try:
        out = subprocess.check_output(
            [
                exe,
                "--query-gpu=index,name,memory.total,memory.used,memory.free,utilization.gpu,temperature.gpu",
                "--format=csv,noheader,nounits",
            ],
            text=True,
            stderr=subprocess.DEVNULL,
            timeout=10,
            **_subprocess_kwargs(),
        )
    except (subprocess.SubprocessError, OSError):
        return [], False

    gpus: list[GpuInfo] = []
    for line in out.splitlines():
        line = line.strip()
        if not line:
            continue
        parts = [p.strip() for p in line.split(",")]
        if len(parts) < 5:
            continue

        def num(i: int) -> int:
            try:
                return int(float(parts[i].split()[0]))
            except (ValueError, IndexError):
                return 0

        def opt_float(i: int) -> float | None:
            if i >= len(parts):
                return None
            try:
                return float(parts[i].split()[0])
            except (ValueError, IndexError):
                return None

        gpus.append(
            GpuInfo(
                index=num(0),
                name=parts[1],
                vram_total_mb=num(2),
                vram_used_mb=num(3),
                vram_free_mb=num(4),
                utilization_pct=opt_float(5),
                temperature_c=opt_float(6),
            )
        )
    return gpus, True


def probe() -> HardwareSnapshot:
    gpus, smi_ok = _run_nvidia_smi()
    snap = HardwareSnapshot(gpus=gpus, primary_gpu=gpus[0] if gpus else None, nvidia_smi_ok=smi_ok)

    if psutil:
        vm = psutil.virtual_memory()
        snap.ram_total_mb = int(vm.total / (1024 * 1024))
        snap.ram_used_mb = int(vm.used / (1024 * 1024))
        snap.ram_available_mb = int(vm.available / (1024 * 1024))
        snap.cpu_count = psutil.cpu_count(logical=True) or 1
        snap.cpu_usage_pct = float(psutil.cpu_percent(interval=0.1))
        disk = psutil.disk_usage(str(Path.home().anchor or "/"))
        snap.disk_free_mb = int(disk.free / (1024 * 1024))
        snap.disk_total_mb = int(disk.total / (1024 * 1024))
        snap.disk_path = str(Path.home().anchor or "/")
    else:
        snap.notes.append("psutil not installed; limited host metrics.")

    if not smi_ok:
        snap.notes.append(
            "NVIDIA management tools were not found. Install a current NVIDIA driver to enable GPU training."
        )

    try:
        import torch  # type: ignore

        snap.cuda_ready = bool(torch.cuda.is_available())
    except Exception:
        snap.cuda_ready = None

    return snap


def guess_params_b(model_path: str) -> float:
    lower = model_path.lower()
    for key, val in [
        ("70b", 70.0),
        ("34b", 34.0),
        ("32b", 32.0),
        ("14b", 14.0),
        ("13b", 13.0),
        ("12b", 12.0),
        ("9b", 9.0),
        ("8b", 8.0),
        ("7b", 7.0),
        ("3b", 3.0),
        ("1b", 1.0),
        ("0.5b", 0.5),
    ]:
        if key in lower:
            return val
    return 7.0


def _fmt_mb(mb: int) -> str:
    if mb >= 1024:
        return f"{mb / 1024:.1f} GB"
    return f"{mb} MB"


def plan_training(
    model_path: str,
    power_mode: str = "balanced",
    material_bytes: int = 0,
    trainable: bool = True,
    estimated_params_b: float | None = None,
) -> TrainingPlan:
    """Build a plan from *this machine's* free resources."""
    hw = probe()
    hard: list[str] = []
    soft: list[str] = []

    free_vram = hw.primary_gpu.vram_free_mb if hw.primary_gpu else 0
    total_vram = hw.primary_gpu.vram_total_mb if hw.primary_gpu else 0

    if not hw.primary_gpu:
        hard.append(
            "No NVIDIA GPU was detected. Training requires an NVIDIA GPU with a current driver."
        )
    if not trainable:
        hard.append(
            "The selected model format cannot be trained directly. Choose a Hugging Face base model folder."
        )

    params_b = estimated_params_b if estimated_params_b is not None else guess_params_b(model_path)
    base_need = int(params_b * 1000 * 0.65 + 1500)

    mode = (power_mode or "balanced").lower()
    if mode == "gentle":
        seq, batch, accum, rank, steps = 1024, 1, 8, 8, 80
        load_mode = "qlora_4bit"
        est = base_need
    elif mode == "faster":
        seq = 4096 if free_vram >= 20000 else 2048
        batch = 2 if free_vram >= 16000 else 1
        accum, rank, steps = 4, 32, 150
        load_mode = "lora_16bit" if free_vram >= 22000 else "qlora_4bit"
        est = int(base_need * 1.25)
    else:
        mode = "balanced"
        seq, batch, accum, rank, steps = 2048, 1, 4, 16, 100
        load_mode = "qlora_4bit"
        est = int(base_need * 1.05)

    # Cap context further on very small free VRAM
    if free_vram and free_vram < 8000:
        seq = min(seq, 1024)
        batch = 1
        load_mode = "qlora_4bit"

    reserve = max(int(total_vram * 0.12), 1024)
    if free_vram > 0 and est + reserve > free_vram:
        if free_vram < 3500:
            hard.append(
                f"Only {_fmt_mb(free_vram)} of GPU memory is free. "
                "Close other GPU apps (e.g. LM Studio with a model loaded), wait a few seconds, then recheck."
            )
        else:
            soft.append(
                f"Free GPU memory is a bit tight ({_fmt_mb(free_vram)} free; plan estimates ~{_fmt_mb(est)}). "
                "You can still start, or free more memory first."
            )

    if hw.ram_available_mb and hw.ram_available_mb < 4096:
        soft.append("System memory is low. Close large applications before training.")

    if material_bytes < 1500:
        soft.append("Very little training material was provided. Results may be weak.")

    if hw.disk_free_mb is not None:
        if hw.disk_free_mb < 15_000:
            hard.append(
                "Less than about 15 GB free disk space remains. Free space before training and export."
            )
        elif hw.disk_free_mb < 40_000:
            soft.append("Disk space is getting low. Exports and checkpoints need room.")

    summary = (
        f"Plan for ~{params_b:.1f}B-class model using {load_mode} on "
        f"{_fmt_mb(free_vram)} free GPU memory ({_fmt_mb(total_vram)} total). Mode: {mode}."
    )

    return TrainingPlan(
        power_mode=mode,
        load_mode=load_mode,
        max_seq_length=seq,
        batch_size=batch,
        grad_accum=accum,
        lora_rank=rank,
        max_steps=steps,
        estimated_vram_mb=est,
        hard_blocks=hard,
        soft_warnings=soft,
        summary=summary,
    )


def to_jsonable(obj: Any) -> Any:
    if hasattr(obj, "__dataclass_fields__"):
        return asdict(obj)
    return obj


PARAM_RE = re.compile(r"(\d+(?:\.\d+)?)b", re.I)

File: src-tauri/src/hw.rs
use crate::win_cmd;
use serde::Serialize;
use std::sync::Mutex;
use std::time::Duration;
use sysinfo::System;

#[derive(Debug, Clone, Serialize)]
pub struct GpuInfo {
    pub index: u32,
    pub name: String,
    pub vram_total_mb: u64,
    pub vram_used_mb: u64,
    pub vram_free_mb: u64,
    pub utilization_pct: Option<f32>,
    pub temperature_c: Option<f32>,
}

#[derive(Debug, Clone, Serialize)]
pub struct HardwareSnapshot {
    pub gpus: Vec<GpuInfo>,
    pub primary_gpu: Option<GpuInfo>,
    pub ram_total_mb: u64,
    pub ram_used_mb: u64,
    pub ram_available_mb: u64,
    pub cpu_count: usize,
    pub cpu_usage_pct: Option<f32>,
    pub disk_free_mb: Option<u64>,
    pub disk_total_mb: Option<u64>,
    pub disk_path: Option<String>,
    pub nvidia_smi_ok: bool,
    pub cuda_ready: Option<bool>,
    pub notes: Vec<String>,
}

/// Kept alive so CPU % is measured between refreshes (sysinfo needs two samples).
static SYS: Mutex<Option<System>> = Mutex::new(None);

/// Probe host resources. Never assumes a specific GPU model or VRAM size.
pub fn snapshot() -> HardwareSnapshot {
    let mut guard = SYS.lock().unwrap_or_else(|e| e.into_inner());
    if guard.is_none() {
        let mut s = System::new();
        s.refresh_memory();
        s.refresh_cpu_usage();
        // First CPU sample is meaningless; take a short second sample.
        std::thread::sleep(Duration::from_millis(200));
        s.refresh_cpu_usage();
        *guard = Some(s);
    } else if let Some(sys) = guard.as_mut() {
        sys.refresh_memory();
        sys.refresh_cpu_usage();
    }

    let sys = guard.as_ref().unwrap();
    let ram_total_mb = sys.total_memory() / (1024 * 1024);
    let ram_used_mb = sys.used_memory() / (1024 * 1024);
    let ram_available_mb = sys.available_memory() / (1024 * 1024);
    let cpu_count = sys.cpus().len().max(1);
    // Clamp silly first-read values
    let raw_cpu = sys.global_cpu_usage();
    let cpu_usage_pct = if raw_cpu.is_finite() {
        Some(raw_cpu.clamp(0.0, 100.0))
    } else {
        None
    };

    let (gpus, nvidia_smi_ok) = query_nvidia_gpus();
    let primary_gpu = gpus.first().cloned();

    let mut notes = Vec::new();
    if !nvidia_smi_ok {
        notes.push(
            "A supported NVIDIA graphics driver was not found. Training needs one.".into(),
        );
    } else if gpus.is_empty() {
        notes.push("No NVIDIA graphics card was reported.".into());
    }

    let (disk_free_mb, disk_total_mb, disk_path) = disk_for_data_dir();

    HardwareSnapshot {
        gpus,
        primary_gpu,
        ram_total_mb,
        ram_used_mb,
        ram_available_mb,
        cpu_count,
        cpu_usage_pct,
        disk_free_mb,
        disk_total_mb,
        disk_path,
        nvidia_smi_ok,
        cuda_ready: None,
        notes,
    }
}

fn query_nvidia_gpus() -> (Vec<GpuInfo>, bool) {
    // Must use CREATE_NO_WINDOW — nvidia-smi is a console app and will flash windows otherwise.
    let output = win_cmd::command("nvidia-smi")
        .args([
            "--query-gpu=index,name,memory.total,memory.used,memory.free,utilization.gpu,temperature.gpu",
            "--format=csv,noheader,nounits",
        ])
        .output();

    let Ok(out) = output else {
        return (Vec::new(), false);
    };
    if !out.status.success() {
        return (Vec::new(), false);
    }

    let text = String::from_utf8_lossy(&out.stdout);
    let mut gpus = Vec::new();
    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let parts: Vec<&str> = line.split(',').map(|s| s.trim()).collect();
        if parts.len() < 5 {
            continue;
        }
        let index = parts[0].parse().unwrap_or(0);
        let name = parts[1].to_string();
        let total = parse_u64(parts[2]);
        let used = parse_u64(parts[3]);
        let free = parse_u64(parts[4]);
        let util = parts.get(5).and_then(|s| s.parse().ok());
        let temp = parts.get(6).and_then(|s| s.parse().ok());
        gpus.push(GpuInfo {
            index,
            name,
            vram_total_mb: total,
            vram_used_mb: used,
            vram_free_mb: free,
            utilization_pct: util,
            temperature_c: temp,
        });
    }
    (gpus, true)
}

fn parse_u64(s: &str) -> u64 {
    s.split_whitespace()
        .next()
        .and_then(|x| x.parse().ok())
        .unwrap_or(0)
}

fn disk_for_data_dir() -> (Option<u64>, Option<u64>, Option<String>) {
    let path = dirs::data_local_dir().unwrap_or_else(|| std::path::PathBuf::from("."));
    let path_str = path.display().to_string();

    let disks = sysinfo::Disks::new_with_refreshed_list();
    let mut best: Option<(u64, u64, String)> = None;
    for disk in disks.list() {
        let mount = disk.mount_point().to_string_lossy().to_string();
        let matches = path_str
            .get(..2)
            .map(|prefix| mount.to_uppercase().starts_with(&prefix.to_uppercase()))
            .unwrap_or(false);
        let free = disk.available_space() / (1024 * 1024);
        let total = disk.total_space() / (1024 * 1024);
        if matches {
            return (Some(free), Some(total), Some(mount));
        }
        if best.as_ref().map(|(f, _, _)| free > *f).unwrap_or(true) {
            best = Some((free, total, mount));
        }
    }
    if let Some((f, t, m)) = best {
        (Some(f), Some(t), Some(m))
    } else {
        (None, None, Some(path_str))
    }
}

/// Adaptive training plan from live hardware — not tied to a fixed GPU SKU.
#[derive(Debug, Clone, Serialize)]
pub struct TrainingPlan {
    pub power_mode: String,
    pub load_mode: String,
    pub max_seq_length: u32,
    pub batch_size: u32,
    pub grad_accum: u32,
    pub lora_rank: u32,
    pub max_steps: u32,
    pub estimated_vram_mb: u64,
    pub hard_blocks: Vec<String>,
    pub soft_warnings: Vec<String>,
    pub summary: String,
}

pub fn plan_training(
    model_path: &str,
    power_mode: &str,
    material_bytes: u64,
    _skill_name: &str,
    model_trainable: bool,
    estimated_params_b: Option<f32>,
) -> TrainingPlan {
    let hw = snapshot();
    let mut hard_blocks = Vec::new();
    let mut soft_warnings = Vec::new();

    let free_vram = hw
        .primary_gpu
        .as_ref()
        .map(|g| g.vram_free_mb)
        .unwrap_or(0);
    let total_vram = hw
        .primary_gpu
        .as_ref()
        .map(|g| g.vram_total_mb)
        .unwrap_or(0);

    if hw.primary_gpu.is_none() {
        hard_blocks.push(
            "No NVIDIA graphics card was found. Teaching a model needs one.".into(),
        );
    }

    if !model_trainable {
        hard_blocks.push(
            "This model is not in a form ModelShaper can teach. Choose a full model folder (not only a single chat file).".into(),
        );
    }

    let params_b = estimated_params_b.unwrap_or_else(|| guess_params_from_path(model_path));

    let base_need = ((params_b * 1000.0) * 0.65 + 1500.0) as u64;
    let (seq, batch, accum, rank, steps, load_mode, est_vram) = match power_mode {
        "gentle" => (
            1024u32,
            1u32,
            8u32,
            8u32,
            80u32,
            "qlora_4bit",
            base_need,
        ),
        "faster" => (
            if free_vram >= 20000 { 4096 } else { 2048 },
            if free_vram >= 16000 { 2 } else { 1 },
            4,
            32,
            150,
            if free_vram >= 22000 {
                "lora_16bit"
            } else {
                "qlora_4bit"
            },
            (base_need as f64 * 1.25) as u64,
        ),
        _ => (
            2048,
            1,
            4,
            16,
            100,
            "qlora_4bit",
            (base_need as f64 * 1.05) as u64,
        ),
    };

    // Soft reserve: warn early, only hard-block when free VRAM is truly too low.
    // (Closing LM Studio after a warning should allow training once memory is free again.)
    let reserve = ((total_vram as f64) * 0.12).max(1024.0) as u64;
    if free_vram > 0 && est_vram + reserve > free_vram {
        if free_vram < 3500 {
            hard_blocks.push(format!(
                "Only {} of graphics memory is free right now. Close other heavy apps (for example LM Studio with a model loaded), wait a few seconds, then use Recheck this PC.",
                format_mb(free_vram)
            ));
        } else {
            soft_warnings.push(format!(
                "Graphics memory is a bit tight ({} free; plan wants about {}). You can still start, or free more memory first.",
                format_mb(free_vram),
                format_mb(est_vram)
            ));
        }
    }

    if hw.ram_available_mb < 4096 {
        soft_warnings.push(
            "System memory is low. Close large applications before starting.".into(),
        );
    }

    if material_bytes < 1500 {
        soft_warnings.push(
            "Very little material was added. Results may be weak.".into(),
        );
    }

    if let Some(free_disk) = hw.disk_free_mb {
        if free_disk < 15_000 {
            hard_blocks.push(
                "There is not enough free disk space (need roughly 15 GB free).".into(),
            );
        } else if free_disk < 40_000 {
            soft_warnings.push(
                "Disk space is getting low. Exports need room.".into(),
            );
        }
    }

    let summary = format!(
        "Plan for about a {:.1}B-size model. Graphics free memory: {}. Mode: {}.",
        params_b,
        format_mb(free_vram),
        power_mode
    );

    TrainingPlan {
        power_mode: power_mode.to_string(),
        load_mode: load_mode.to_string(),
        max_seq_length: seq,
        batch_size: batch,
        grad_accum: accum,
        lora_rank: rank,
        max_steps: steps,
        estimated_vram_mb: est_vram,
        hard_blocks,
        soft_warnings,
        summary,
    }
}

fn guess_params_from_path(path: &str) -> f32 {
    let lower = path.to_lowercase();
    for (k, v) in [
        ("70b", 70.0),
        ("65b", 65.0),
        ("34b", 34.0),
        ("33b", 33.0),
        ("32b", 32.0),
        ("27b", 27.0),
        ("22b", 22.0),
        ("14b", 14.0),
        ("13b", 13.0),
        ("12b", 12.0),
        ("9b", 9.0),
        ("8b", 8.0),
        ("7b", 7.0),
        ("3b", 3.0),
        ("1b", 1.0),
        ("0.5b", 0.5),
    ] {
        if lower.contains(k) {
            return v;
        }
    }
    7.0
}

fn format_mb(mb: u64) -> String {
    if mb >= 1024 {
        format!("{:.1} GB", mb as f64 / 1024.0)
    } else {
        format!("{} MB", mb)
    }
}
